### PR TITLE
Implement more Variable Coders

### PR DESCRIPTION
**Pull request by kmuehlbauer**
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

This PR implements remaining functionality in `encode_cf_variable`/`decode_cf_variable` as `VariableCoders` (as left as todo by @shoyer). It adapts the needed tests and adds some typing (thanks @Illviljan). Otherwise there are no changes in function to code.

It splits out of #7654 which turns out to be hard to follow (at least for me).

### CODE REVIEW

**Review by Illviljan — APPROVED** [review]

File: xarray/coding/variables.py [hunk 1]
```diff
@@ -159,30 +224,34 @@ def encode(self, variable: Variable, name: T_Name = None):
...
 
-        if (
-            fv is not None
-            and mv is not None
-            and not duck_array_ops.allclose_or_equiv(fv, mv)
-        ):
-            raise ValueError(
-                f"Variable {name!r} has conflicting _FillValue ({fv}) and missing_value ({mv}). Cannot encode data."
-            )
+        if fv is not None or mv is not None:
```
Comment (Illviljan) [inline xarray/coding/variables.py:227]:
```suggestion
        fv_exists = fv is not None
        mv_exists = mv is not None
        if fv_exists or mv_exists:
```
A lot of the same checks over and over again, can make a variable instead and reuse that.
You can replace more below.

Comment (kmuehlbauer) [reply xarray/coding/variables.py:227]: Thanks I've already thought about how to make this more readable. Just out of interest @Illviljan, is this change more for readability or does it also have an impact on performance?

Comment (Illviljan) [reply xarray/coding/variables.py:227]: Was mainly a micro-optimization to me, nothing major. From like 4x-"fx is not None" operations to 1. 
I just think it's good practice to avoid duplicating operations.

File: xarray/coding/variables.py [hunk 2]
```diff
@@ -78,6 +78,71 @@ def __repr__(self) -> str:
         )
 
 
+class NativeEndiannessArray(indexing.ExplicitlyIndexedNDArrayMixin):
```
Comment (dcherian) [inline xarray/coding/variables.py:81]: It's good to mention that things have just been moved around without modifications, if that is the case. If not, its good to point out the differences for the reviewer.

File: xarray/coding/variables.py [hunk 3]
```diff
@@ -159,30 +224,34 @@ def encode(self, variable: Variable, name: T_Name = None):
...
+                encoding["missing_value"] = dtype.type(mv)
+                fill_value = pop_to(encoding, attrs, "missing_value", name=name)
+                if not pd.isnull(fill_value) and fv is None:
+                    data = duck_array_ops.fillna(data, fill_value)
 
-        return Variable(dims, data, attrs, encoding, fastpath=True)
+            return Variable(dims, data, attrs, encoding, fastpath=True)
+
+        else:
+            return variable
```
Comment (dcherian) [inline xarray/coding/variables.py:254]: Seems like this diff would be smaller, or at least easier to read, if we `return variable` early when `fv is None and mv is None`

File: xarray/coding/variables.py [hunk 4]
```diff
@@ -349,3 +418,99 @@ def decode(self, variable: Variable, name: T_Name = None) -> Variable:
...
+            attrs["_FillValue"] = variable.dtype.type(np.nan)
+            return Variable(dims, data, attrs, encoding, fastpath=True)
+        else:
+            return variable
+
+    def decode(self, variable: Variable, name: T_Name = None) -> Variable:
+        raise NotImplementedError()
+
+
+class BooleanCoder(VariableCoder):
```
Comment (dcherian) [inline xarray/coding/variables.py:443]: It seems like code was only moved around, no functional changes?

File: xarray/conventions.py [hunk 5]
```diff
@@ -389,19 +276,15 @@ def decode_cf_variable(
...
-    # TODO(shoyer): convert everything below to use coders
+    if decode_endianness and not var.dtype.isnative:
+        var = variables.EndianCoder().decode(var)
+        original_dtype = var.dtype
 
-    if decode_endianness and not data.dtype.isnative:
-        # do this last, so it's only done if we didn't already unmask/scale
-        data = NativeEndiannessArray(data)
-        original_dtype = data.dtype
+    var = variables.BooleanCoder().decode(var)
```
Comment (dcherian) [inline xarray/conventions.py:283]: Can you add a comment for why the order was changed here? It now comes before `encoding.setdefault("dtype", original_dtype)`

**Review by dcherian — COMMENTED** [review]
Thanks @kmuehlbauer ! Good stuff!

File: xarray/coding/variables.py (hunk 2 above)
Comment (kmuehlbauer) [reply xarray/coding/variables.py:81]: Yes will do, I'll put a comment in the code.

Comment (dcherian) [reply xarray/coding/variables.py:81]: Sorry for the confusion. A GitHub comment is enough, it's just for the review.

Comment (kmuehlbauer) [reply xarray/coding/variables.py:81]: These were just moved from conventions.py with typing added.

File: xarray/coding/variables.py (hunk 4 above)
Comment (kmuehlbauer) [reply xarray/coding/variables.py:443]: Just moved around. I did not want to make functional changes in this PR.

File: xarray/conventions.py (hunk 5 above)
Comment (kmuehlbauer) [reply xarray/conventions.py:283]: I need to check, might be by purpose. I'll come back later with details.

File: xarray/coding/variables.py (hunk 3 above)
Comment (kmuehlbauer) [reply xarray/coding/variables.py:254]: Yes, this might be easier to find then  as below the long code block.

File: xarray/conventions.py (hunk 5 above)
Comment (kmuehlbauer) [reply xarray/conventions.py:283]: `BooleanEncoder ` doesn't need or change contents of `.encoding`. So the order doesn't matter here. In the light of this I've moved this before the unpacking to only have to unpack once before applying `original_dtype` and `LazyIndexedArray`.

File: xarray/coding/variables.py (hunk 1 above)
Comment (kmuehlbauer) [reply xarray/coding/variables.py]: I've applied your suggestion, @Illviljan together with @dcherian's.

File: xarray/coding/variables.py (hunk 3 above)
Comment (kmuehlbauer) [reply xarray/coding/variables.py]: - [x] Done

**Comment by kmuehlbauer** [comment]
Now, this is interesting! It looks like those FillValue issues are following me. What did change that this now materializes here, all of a sudden.

**Comment by kmuehlbauer** [comment]
This looks like it is ready to go. This will surely help further refactoring `encode_cf_variable`/`decode_cf_variable`. At least while working on it I spotted several locations where inconsistencies can be ironed out. A neat mostly flaw-free encoding/decoding is needed especially with regard to #6323.

**Comment by dcherian** [comment]
Thanks @kmuehlbauer

**Review by dcherian — APPROVED** [review]
